tests/pipeline_test.py

```diff
@@ -1208,3 +1208,36 @@ def process(
 def test_html_repr() -> None:
     pipeline = sl.Pipeline([make_int], params={float: 5.0})
     assert isinstance(pipeline._repr_html_(), str)
+
+
+def test_does_not_allow_type_argument_outside_of_constraints() -> None:
+    T = TypeVar('T', int, float)
+
+    @dataclass
+    class M(Generic[T]):
+        value: T
+
+    S = TypeVar('S', M[int], M[float])
+
+    @dataclass
+    class N(Generic[S]):
+        value: S
+
+    def p1(value: T) -> M[T]:
+        return M(value)
+
+    def p2(value: S) -> N[S]:
+        return N(value)
+
+    pipeline = sl.Pipeline((p1, p2))
+    pipeline[str] = 'abc'
+    pipeline[int] = 123
+
+    pipeline.get(M[int])
+    pipeline.get(N[M[int]])
```

Comment:
I think I had written a less contrived example in the related issue, can we add that as an additional simpler test?

Reply:
I copied the test you wrote in the issue, the only difference is that I added the nested case. Would it be better if I split the not-nested and the nested case into two tests?

Reply:
Well, it does test one thing, but in two different cases, which is common pratice...

But I'll split it into two separate tests if you think that's better in this case.